akka-http-core/src/main/scala/akka/http/impl/engine/http2/PingTimeoutBidi.scala

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.http2
+
+import akka.NotUsed
+import akka.http.impl.engine.http2.FrameEvent.PingFrame
+import akka.http.impl.engine.ws.Protocol.Opcode.Ping
+import akka.http.impl.engine.ws.Protocol.Opcode.Ping
+import akka.http.impl.engine.ws.Protocol.Opcode.Ping
+import akka.http.impl.engine.ws.Protocol.Opcode.Ping
+import akka.http.impl.engine.ws.Protocol.Opcode.Ping
+import akka.http.scaladsl.settings.Http2ClientSettings
+import akka.stream
+import akka.stream.Attributes
+import akka.stream.BidiShape
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.scaladsl.BidiFlow
+import akka.stream.stage.GraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+import akka.stream.stage.TimerGraphStageLogic
+import akka.util.ByteString
+
+import scala.concurrent.duration.DurationInt
+
+object PingTimeoutBidi extends {
+
+  def apply(settings: Http2ClientSettings): BidiFlow[FrameEvent, FrameEvent, FrameEvent, FrameEvent, NotUsed] =
+    BidiFlow.fromGraph(new PingTimeoutStage(settings))
+
+  private val tick = "tick"
+  private val Ping = PingFrame(false, ByteString("abcdefgh")) // FIXME should we have unique ping payloads so we can identify?
+
+  // FIXME ignore ping/pong for the connection idle timeout? (by merging that logic into this stage perhaps?)
```

Comment:
I'm not sure that is necessary: since missed pings would lead to a disconnect as well, perhaps it's no problem that the idle timeout is suppressed when there's pings?

Reply:
Even if we don't implement the proposed logic from https://github.com/grpc/grpc/blob/master/doc/keepalive.md to not ping when there's no requests in flight, this would allow users to choose between 'active' pinging and a 'passive' idle timeout.